Remove incomplete error state in transaction_new

CiaranWoodward opened the pull request:
Discovered what looked like a potential memory leak in the 'error' state of transaction_new. However it wasn't actually possible to hit this state without a programmer error (utils_intToText will never fail with a length 6 buffer and a uint16_t int)... but the state did exist and didn't properly clear up. One option was to correct it, so that it cleaned up properly:

```C
error:
    LOG("Exiting on failure");
    if(transacP->message) 
    {
        coap_free_header(transacP->message);
        lwm2m_free(transacP->message);
    }
    lwm2m_free(transacP);
    return NULL;
```

But I figured it was cleaner to just remove the unused & incomplete code.

sbertin-telular commented:
I'm torn between approving of removing code paths that can't be executed and disapproving of the removal of error checking that could catch future problems. I'm leaning toward approval. Any other opinions?

sbernard31 commented:
Classic dilemma. 

I guess this is a bet based on : 
- cost of checking
- risk prevented by this check

My 2 cts in this case : I feel that cost checks is very low but in a meantime I don't to see which kind of risk it could concretely prevent.(but maybe I just missed it)
So I would rather go for removing. (Clearly not strong opinion)

sbernard31 commented:
@qleisan  any opinion about this ?

qleisan commented:
The potential malloc runtime error is handled and that is great. Potential runtime errors should always be checked and handled. 
Removal of the check for `result == 0` is questionable I think. Programming errors cannot / should not always be checked for but in this case it is a) a pattern used on many other places when the function is called b) a call to a function that explicitly returns 0 in several places. If the `result == 0` check is removed I think a comment should be added to help others feel good about this when reading the code, It does get cleaner without the check...

sbernard31 commented:
I'm not sure we helped @CiaranWoodward to know what he should do :sweat_smile: 

My understanding :  I feel that the team is not totally comfortable with the idea of removing the check
But  this sounds clearly not to be a strong opinion.

So @CiaranWoodward, I guess this is up to you to decide what to do : 
1. either keep the error check
2. or remove it (but in this case a comment would be appreciated)

@sbertin-telular, @qleisan  do not hesitate to correct me if I misunderstood the situation.

CiaranWoodward commented:
Hi, apologies for being pretty silent here. I'll take another look and see if I can find a more graceful fix - can always fall back to the snippet I posted in the initial comment.

What is this project's stance on `assert``s? I noticed that there weren't any in the codebase, but they can be a good way to catch coding errors. Any thoughts?

CiaranWoodward commented:
I looked into it and seeing as it isn't that big of a deal for code size or stability, I reverted 25e495b5e76b1eec6b366e34938dc3a5abebfa04 and replaced it with the snipped from the comment. It does still seem out of place to me to catch an error that can never happen, but can chalk it up to different style. It may be worth looking into solutions like ``assert``, that can signal programmer intent and just crash the program upon error.

The new commit has been pushed - hopefully that solves the issue!

sbernard31 commented:
@CiaranWoodward thx for your contribution :pray: 

@tuve, @qleisan , @sbertin-telular  do not hesitate to share your point of view about `assert` usage. (Just in case I added a comment about this in #513)
